### PR TITLE
fix(core): global listeners not being bound on non-node host elements

### LINKED ISSUE #14191: HostListener doesn't work on structural directives (with global events)
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Placing a host listener on a structural directive doesn't work (no error, nothing).

**Expected behavior**
We should be able to use HostListener. Or it should throw a meaningful error if it's expected behavior.

**Minimal reproduction of the problem**

http://plnkr.co/edit/VOmMkLWSEVONy3cKXZJq?p=preview

**Please tell us about your environment:**

* **Angular version:** 2.4.4

* **Browser:** Chrome 54

 
* **Language:** TypeScript

### PR DESCRIPTION
**Pull request by crisbeto**
We skip event listeners on non-element host nodes (e.g. `ng-container` or `ng-element`), because they don't map to a DOM node so there's nothing to bind the event to. The problem is that this also prevents listeners bound to global targets from being bound.

These changes add an extra condition to allow for the event to be bound if it has a custom event target resolver.

Fixes #14191.

### CODE REVIEW

File: packages/core/test/acceptance/listener_spec.ts [hunk 1]
```diff
@@ -397,6 +398,71 @@ describe('event listeners', () => {
...
+        class AddGlobalListener {
+          @HostListener('document:click')
+          handleClick() {
+            clicks++;
+          }
+        }
+
+        @Component({
+          template: `
+            <ng-template #template add-global-listener>
```
Comment (splincode) [inline packages/core/test/acceptance/listener_spec.ts:446]: what happens if someone writes the code like this

```html
 
            <ng-template #template add-global-listener>
              <button>Click me!</button>
            </ng-template>
            <ng-container [ngTemplateOutlet]="template" add-global-listener></ng-container>
          
```

Comment (splincode) [reply packages/core/test/acceptance/listener_spec.ts:446]: add test please for it case

Comment (crisbeto) [reply packages/core/test/acceptance/listener_spec.ts:446]: This is basically the same use case as the `ng-container` one above, except that it'll trigger 2 listeners. I'm not sure how much value it'll add.

File: packages/core/test/acceptance/listener_spec.ts [hunk 2]
```diff
@@ -397,6 +398,71 @@ describe('event listeners', () => {
...
+        const fixture = TestBed.createComponent(MyComp);
+        fixture.detectChanges();
+        const button = fixture.nativeElement.querySelector('button');
+        button.click();
+        fixture.detectChanges();
+        expect(clicks).toBe(1);
+      });
+
+  onlyInIvy('issue has only been resolved for Ivy')
+      .it('should bind global event listeners on an ng-template directive host', () => {
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/listener_spec.ts:433]: I'd propose adding one more case where we use similar directive as a structural one:

```
 <div *add-global-listener>...</div>
```

That would de-sugar to `<ng-template>` with extra attributes (which you have covered below), but such a test case can be helpful to document the behavior (and make sure it doesn't regress in the future).

**Review by AndrewKushnir — APPROVED** [review]
@crisbeto looks great, thanks for the fix 👍 

I've just left a couple comments on updating the comment and adding a test case with a structural directive.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:372954623:BASE:372954637:1620666600830:9ed11499).

**Comment by crisbeto** [comment]
The feedback has been addressed.

**Comment by AndrewKushnir** [comment]
@crisbeto FYI, presubmits went well for the changes in this PR.

**Comment by AndrewKushnir** [comment]
@crisbeto one quick comment: I'm wondering if we should also look into better detection of the case where host listeners are added for the non-node elements and provide some helpful logs (in the console)? This is outside of this ticket, but it just came to mind while reviewing this change. What's the current behavior (do we just silently ignore such cases)?

**Comment by crisbeto** [comment]
Yes, the current behavior is to silently ignore them. I believe that it might've been modeled after ViewEngine. A warning could be useful, but I don't know whether it would be actionable by the user. E.g. there might be a directive that can be set both on a regular node and an inert one and which has some optional event listener.

**Comment by crisbeto** [comment]
Based on our discussion yesterday and the fact that the presubmits for this went well earlier, I'm marking it as ready to merge.

**Pull request closed without merging**